client-state-machine.go

```diff
@@ -59,12 +59,13 @@ type ClientStateStart struct {
 	helloRetryRequest *HandshakeMessage
 }
 
-func (state ClientStateStart) Next(hm *HandshakeMessage) (HandshakeState, []HandshakeAction, Alert) {
-	if hm != nil {
-		logf(logTypeHandshake, "[ClientStateStart] Unexpected non-nil message")
-		return nil, nil, AlertUnexpectedMessage
-	}
+var _ HandshakeState = &ClientStateStart{}
```

Comment:
What does this line do?  It seems like a noop.  I think you can safely remove this and the ones below it.

Reply:
Yes, it's a noop. It's a check that the struct satisfies the interface. I like this pattern because you'll get the compile error right there if it doesn't.